src/dragonball/src/device_manager/blk_dev_mgr.rs

```diff
@@ -404,9 +406,30 @@ impl BlockDeviceMgr {
                             BlockDeviceError::DeviceManager(e)
                         })
                     }
+                    #[cfg(feature = "vhost-user-blk")]
                     BlockDeviceType::Spool | BlockDeviceType::Spdk => {
-                        // TBD
-                        todo!()
+                        let device = Self::create_vhost_usr_device(&config, &mut ctx)
+                            .map_err(BlockDeviceError::Virtio)?;
+                        let dev = DeviceManager::create_mmio_virtio_device(
+                            device,
+                            &mut ctx,
+                            config.use_shared_irq.unwrap_or(self.use_shared_irq),
+                            config.use_generic_irq.unwrap_or(USE_GENERIC_IRQ),
+                        )
+                        .map_err(BlockDeviceError::DeviceManager)?;
+                        self.update_device_by_index(index, Arc::clone(&dev))?;
```

Comment:
I can't find the implementation of the `update_device_by_index`. So, Is there no need to do map_err() here ?

Reply:
You can find the implementation of this function in L784.

Reply:
IMO, based on the  implementation of `update_device_by_index`, if error occurs, it seems that there's not enough error message dumpped for users, right?